signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+import os
+import json
+import errno
+import uuid
+from copy import copy
+from contextlib import contextmanager
+from collections.abc import Mapping
+from collections.abc import MutableMapping
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # create Collection
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+        @classmethod
+        def __instancecheck__(cls, instance):
+            for parent in cls.__mro__:
+                if not isinstance(instance, parent):
+                    return False
+            return True
```

Comment:
For posterity, this code is adapted almost verbatim from the Python library itself (https://github.com/python/cpython/blob/master/Lib/_collections_abc.py). @vishav1771 we'll need to be careful and double-check that this implementation is valid. The Python documentation has conflicting information and I'm not 100% sure I gave you good advice regarding the implementation of `__instancecheck__` and `__subclasshook__`. The [Python Data Model](https://docs.python.org/3/reference/datamodel.html#customizing-instance-and-subclass-checks) says explicitly that these methods must be implemented on the *metaclass*, but the documentation for the [abc module](https://docs.python.org/3/library/abc.html#module-abc) says that they can be implemented as `classmethod`s of subclasses. My interpretation was that the `ABCMeta` class contains the magic ingredients to allow this, but honestly I haven't delved into the code enough to verify this so we'll just want to make sure that all calls to `isinstance` and `issubclass` are tested thoroughly when you write unit tests.

This issue also applies to the `__instancecheck__` that I implemented for `SyncedCollection`, so we'll need to carry over whatever decision we make here to that as well.

Reply:
I tried it. I found out that the `__instancecheck__` method should be defined on _metaclass_. I think in _abc module_ documentation, they are talking about `__subclasshook__` not `__instancecheck__`.